apps/autoscheduler/LoopNest.cpp

```diff
@@ -0,0 +1,1861 @@
+    parallel = n.parallel;
+    vector_dim = n.vector_dim;
+    vectorized_loop_index = n.vectorized_loop_index;
+};
+
+// Hash the loop structure and sizes up to a fixed depth. This is
+// used as the hash function for the coarse-to-fine beam search in
+// the paper.
+void LoopNest::structural_hash(uint64_t &h, int depth) const {
+    if (depth < 0) return;
+
+    // Which Funcs are store_at this level?
+    for (const auto *n : store_at) {
+        hash_combine(h, n->id);
+    }
+
+    hash_combine(h, -1);
+
+    // Which Funcs are compute_at this level?
+    for (const auto &c : children) {
+        hash_combine(h, c->stage->id);
+    }
+
+    // Add a barrier to ensure that moving something from the last
+    // compute_at to the first inlined doesn't result in the same
+    // hash.
+    hash_combine(h, -1);
+
+    // Which Funcs are inlined at this level?
+    for (auto it = inlined.begin(); it != inlined.end(); it++) {
+        hash_combine(h, it.key()->id);
+    }
+
+    hash_combine(h, -1);
+
+    if (depth > 0) {
+        // What are the loop sizes of the children?
+        for (const auto &c : children) {
+            for (int64_t s : c->size) {
+                if (depth == 1) {
+                    // Just take the most significant bit: is it one or not?
+                    s = (s > 1) ? 1 : 0;
+                }
+                hash_combine(h, s);
+            }
+        }
+
+        // Which dimension are we vectorized over?
+        hash_combine(h, vectorized_loop_index);
+    }
+
+    if (depth > 1) {
+        // Descend into children
+        for (const auto &c : children) {
+            c->structural_hash(h, depth - 2);
+        }
+    }
+}
+
+// Compute all the sites of interest for each pipeline stage
+void LoopNest::get_sites(StageMap<Sites> &sites,
+                         const LoopNest *task,
+                         const LoopNest *parent) const {
+    if (!task && !is_root()) {
+        task = this;
+    }
+    for (const auto &c : children) {
+        c->get_sites(sites, task, this);
+    }
+    if (parent && node != parent->node) {
+        auto &s = sites.get_or_create(stage);
+        s.compute = parent;
+        s.produce = this;
+        s.task = task;
+    }
+    for (auto f : store_at) {
+        for (const auto &s : f->stages) {
+            sites.get_or_create(&s).store = this;
+        }
+    }
+    for (auto it = inlined.begin(); it != inlined.end(); it++) {
+        auto &s = sites.get_or_create(&(it.key()->stages[0]));
+        s.inlined = true;
+        s.compute = s.store = s.produce = s.innermost = this;
+        s.task = task;
+    }
+    if (innermost) {
+        sites.get_or_create(stage).innermost = this;
+    }
+}
+
+// A helper for the working_set_at_task feature. Most features are
+// computed in the recursive pass 'compute_features' below, but
+// this one must be done in a second separate recursive pass.
+/*    void set_working_set_at_task_feature(int64_t working_set,
```

Comment:
This looks like a duplicate of the non-commented out function below; please remove

Reply:
deleted